jupyterhub_moss/spawner.py

```diff
@@ -92,29 +91,31 @@ def _validate_partitions(self, proposal):
         help="""Provides information about resources in Slurm cluster.
 
         It will be called with the output of ``slurm_info_cmd`` as argument and should return a
-        dictionary mapping each partition name to resources defined in ``REQUIRED_RESOURCES_COUNTS``
-        and resources used in option_form template.
+        dict mapping each partition name to an instance of a :class:`models.PartitionResources`.
         """,
     ).tag(config=True)
 
     @traitlets.default("slurm_info_resources")
-    def _get_slurm_info_resources_default(self):
+    def _get_slurm_info_resources_default(
+        self,
+    ) -> Callable[[str], dict[str, PartitionResources]]:
         """Returns default for `slurm_info_resources` traitlet."""
         return self._slurm_info_resources
 
-    def _slurm_info_resources(self, slurm_info_out: str) -> Dict[str, dict]:
+    def _slurm_info_resources(
+        self, slurm_info_out: str
+    ) -> dict[str, PartitionResources]:
         """Parses output from Slurm command: sinfo -a --noheader -o '%R %F %c %C %G %m %l'
 
-        Returns information about partition resources listed in ``REQUIRED_RESOURCES_COUNTS``:
-        number of cores, max memory, gpus and resource counts to be shown in table of available resources.
+        Returns information about partition resources to constraint user choice and display available resources.
 
         :param slurm_info_out: Output of slurm_info_cmd
         :rtype: Mapping of partition information:
             {
                 partition: {gpu, max_nprocs, max_ngpus, max_mem, max_runtime, ...},
             }
         """
-        partitions_info = {}
+        partitions_info: dict[str, PartitionResources] = {}
```

Comment:
Explicit typing since it is filled with a model inherited from this, but it need to match the returned type